New Link Format Documentation

lzimm opened the pull request:
this diverges slightly from https://github.com/preludeorg/operator/pull/3629/files and includes some augmentations (the account `organization_id` and `group` specifically) that may be best as future additions, but should capture the spirit of things in a way that makes the future more useful regardless.

(note specifically: ignoring `effects` for now, if we revisit that, it should probably go hand-in-hand with any discussions we have around schema).

privateducky commented on docs/operator/advanced/links.md:
```diff
@@ -0,0 +1,49 @@
+
+#### Attributes
+
+- **unique**: every link has a unique identifier used to differentiate it from other executions, which may even share almost every other attribute in common.
```
should this be unique or id or other?

lzimm replied:
Lol. If that's up for debate I think we should do `id` across the board. Its always made me a little crazy.

lzimm commented on docs/operator/advanced/links.md:
```diff
@@ -0,0 +1,49 @@
+	- `end`: the moment after the command has finished execution and the agent collects your results.
+- **username**: this is the system level account that the agent used to execute the command: whether impersonated or not, this will help you determine, among other things, the privileges and access levels that the agent had access to when executing each command.
+- **hostname**: this is the network hostname of the machine the agent is currently executing on, as of the moment the command ran.
+- **ttp**: the TTP id of the attack being executed
+- **tactic**: the TTP's tactic
+- **technique**: the TTP's technique
```
alternatively:

- **ttp**: the TTP of the attack
    -  `id`: the TTP's id
    - `tactic`: the TTP's tactic
    - `technique`: the TTP's technique

privateducky replied:
yeah, I think this makes the most sense.

let's look to make tactic/technique generic though. ATT&CK will obviously get replaced in the not-distant-future.. using these terms may be fine (just expressing a multi-layer classification) but worth looking at CVE, some NIST-800 (creatively but it'll come up) and other options on classification. I think this may be one of the areas that may be most often modified by the user.. 

one connected question: do you think this info should always map from the TTP itself or are there cases where the TTP is classified one way but the link another? primary case I'm thinking of is around chain metadata and if we'd ever try to classify the chain separately or if it's classification is a sum of the TTPs used.

lzimm replied:
totally agree with the attack thing, that was one of the ones that was grinding me too, but given that most of the other platforms i've looked at all speak attack, that was the first +1 for saying "okay this should work".

the next +1 came from, at the individual ttp level, it may not make sense for some staging command to really correspond to a full CVE, especially when you can compose TTPs in different ways to assemble one of many different CVEs, which is something that's more naturally expressed at the chain level, rather than TTP, IF there's even a CVE in play at all (ie: would a simple chain like "ransom note" correspond to a CVE?).

regardless though, one way we can make this more "future proof" is instead of maintaining everything totally flat inside the `ttp` thing... what if we structured it like this:

- ttp:
  - id
  - attack:
    - tactic
    - technique

which allows us, when we bring in additional taxonomy (at the TTP level), to naturally augment it there, side by side with the `attack` object.

per the chain thing, i was thinking a lot about that too, but i think that one is gonna be really important to keep living outside of the actual links (with reconciliation done through the operation). chains are going to be something that i'm not sure really makes sense to push out into elastic/splunk/etc., because they're either gonna pick up results or not, and as long as they can see things that were meant to go together (by way of a shared operation id), there should be enough there to tell whatever story is relevant, and we can either reconcile operations with chains on OUR side, or through plugins that we might build to keep things in sync.

one thing to note here though, there's a difference between the operation and the origin of the operation. 

one thing i could see as being useful (ie: to track recurring operations that all run the same set of commands on regular intervals, etc.), is tracking the actual source (as a proxy for the "intention") of a set of executions, which could come in valuable in a variety of ways.

in some worlds, the origin would basically be the chain id that the operation was spawned from (which definitely helps make the upstream correlation a bit easier on our side too—a few less joins involved), but it could also be just a one off thing that a user just fired on their own, or even a parent chain spawning child chains, etc. etc, calculating the actual TTPs to run dynamically through some ad-hoc process (although that's clearly a bit speculative).

lends itself to being a bit of an object: ie (`{operation: {id: '', origin: {chain: 'CHAIN_ID_HERE'}}}`), but it might be easier to just assume its always a "chain" for now, as that's really how things are structured for us internally (more or less, in the cases where we'd really care about something like this) anyways (ie: `{operation: {id: ..., chain_id: ...}}`)

privateducky commented on docs/operator/advanced/links.md:
```diff
@@ -0,0 +1,49 @@
+
+- **id**: every link has a unique id used to differentiate it from other executions, which may even share almost every other attribute in common.
+- **operation**: every link belongs to an operation, which is the total set of every TTP, and all of the requests it generates, when when you decide to run a chain.
+- **agent**: we collect the name of the agent that executed the command itself, if you have multiple agents all running on the same host, you can use this value to differentiate between which agent ran what, though you'll have to be careful about naming your agents appropriately to use this value in a way that makes sense.
```
looking at agent as the first layer of "host identification", the others that may be more/less useful:
- Mac address
- Local IP
- Public IP (to help in corporate networks where a public IP indicates a VPN, which would be a useful data cut).
- Host name
- Username (on computer)
- License key (or email tied to license)

thinking of "agent" grouped into this thought process, our current agent name is == the computer user name. for detect agents I wonder if this makes sense in context of the above, or if we should adjust that to be something more tangible... initial thought is that we have a 'identification' object on the link with the options we think are reliable, so people can add/remove from it down the line. 

on the topic of agents, it's a near certainty we end up with different detect agents, based on varying levels of integration points. sort of the detect version of having pneuma/schism/etc. I'm not convinced it is, but I wonder if that level of tracking matters (or if it's ok as an afterthought).

lzimm replied:
so, the statement `our current agent name is == the computer user name` isn't actually true... the agent.name right now is basically the unique agent id that the agent decides to beacon in with.

per the things above, i'm not at all opposed to having the mac address and IPs on the list as well (the bottom 3 area already planned—where the "license key" is basically accomplished by the `account` attribute, which would get saturated going from HQ->outpost).

per the different agents themselves though, my initial thought was that we might want to reserve that kind of reconciliation for some sort of offline process. theoretically we should be able to tie it all together at some other level (ie: through the operations that get scheduled, and the authentication loop that'd need to happen between agent<>hq to schedule those operations in the first place), and in the interest of trying to keep the attributes on the link level tied as closely as possible to the actual execution itself, with all of OUR fingerprints limited to the bare minimum of what we'd need to actually determine what we need to know from there, i'd opt to try to keep any extra agent info out of it.

what might actually be a good idea though is making our agent.name collisions less likely by generating full agent headers, rather than just forwarding through simple agent names on the beacon.

ie:

instead of: `pneuma --name=coldduck` -> `beacon: name: coldduck`
we have more of a useragent header: `pneuma --name=coldduck` -> `beacon: agent: Pneuma/5.0 (Macintosh; Intel Mac OS X 10_15_7): coldduck`

lzimm commented on docs/operator/advanced/links.md:
```diff
@@ -0,0 +1,49 @@
+- **id**: every link has a unique id used to differentiate it from other executions, which may even share almost every other attribute in common.
+- **operation**: every link belongs to an operation, which is the total set of every TTP, and all of the requests it generates, when when you decide to run a chain.
+- **agent**: we collect the name of the agent that executed the command itself, if you have multiple agents all running on the same host, you can use this value to differentiate between which agent ran what, though you'll have to be careful about naming your agents appropriately to use this value in a way that makes sense.
+- **account**: we collect the account details of the user originating each request, this is an object with the following attributes:
```
ohh yeah, if you look at my PR on operator, I'm not (planning on) tracking the actual account until it goes from HQ -> outpost at all.

privateducky replied:
group is interesting. I removed this concept from our schema at the moment. for detect, I'm curious if we need to define it on our side or if it's a param you can set when starting detect agents OR if this isn't our business and we should let the user hook into their LDAP system to resolve accounts (again based on their existing host id/inventory process). will need to mull that over and look at the different ways we can/should hook in with LDAP before this prop gets into the link I think.

lzimm replied:
totally, i just stuffed that in there as a general placeholder for "where this user belongs on YOUR side, not our side", but as long as we have space for that inside of an `account` like object (as opposed to the simple `tag` that we store now), we can enrich that in the ways that make most sense, when we get to that spot of the whole roadmap.

privateducky commented on docs/operator/advanced/links.md:
```diff
@@ -0,0 +1,49 @@
+	- `email`: the users' email.
+	- `organization_id`: the organization id that the user belongs to.
+	- `group`: the users' group inside the organization.
+- **platform**: since each operation can span multiple machines each running a different operating system, it's important to store the platform to help you distinguish your results at the environment level.
```
we stick with higher-level OS' right now because we're not crazy. we'll likely need to do versioning of OS' in detect. I'm not sure how much needs to come from us vs a correlation between their asset inventory system and their users though.. I'm hesitant to put too much related to that in detect results specifically,  as there will be a large tendency to use us for asset inventory.. not a bad thing but if we're aware that'll be a driving interest, we should consider the other parts of asset inventory (i.e., host identification props) that will come up. this ties into patches needed for different OS', which if we kept an open mind, we could make it easy to use detect for asset inventory (but not our purpose)... as in, we'd use a platform syntax that maps directly to patches on Microsoft, CVE, etc. so we could resolve them.

lzimm replied:
yeah i'd personally kinda veery away from that kinda stuff because, as you mentioned, we can correlate it other ways, but furthermore, it starts to bloat things and makes blurry what belongs inside of a link (which we'd use to query that kinda stuff, among many other things) and what doesn't.

privateducky commented on docs/operator/advanced/links.md:
```diff
@@ -0,0 +1,49 @@
+	- `group`: the users' group inside the organization.
+- **platform**: since each operation can span multiple machines each running a different operating system, it's important to store the platform to help you distinguish your results at the environment level.
+- **executor**: since each TTP can be executed using a variety of different executors (ie: bash, powershell, python, etc.), we store the executor selected to run the individual instruction to help you distinguish that too.
+- **payload**: if your TTP required a payload as part of its request, we store the specific payload name and version used.
```
think we need payload specifically? since it's an extension on the TTP anyway, I think we'd just have a link-to-TTP look up on the dashboard, allowing us to pull this info.

I'd contrast that with the classification system used (like ATT&CK) which is likely to be core to how results will be analyzed/cut.

lzimm replied:
the only reason i'd like to include it is that it may make the raw joins inside splunk/elastic/etc. a lot easier, having the payload information inside of the actual link object directly lets you actually build searches that can leverage it directly too. also, since TTPs can change over time, capturing the actual checksum of the payload that ran at the time it ran is another big motivator for persisting it directly.

privateducky commented on docs/operator/advanced/links.md:
```diff
@@ -0,0 +1,49 @@
+	- `output`: the regular output stream (stdout) of the request, which is where data gets parsed out of for downstream commands.
+	- `error`: the error stream (stderr) of the request, which is returned for diagnostic information, with no parsing and data extraction performed.
+- **status**: the status code of the request.
+- **pid**: the process ID of the executed task.
```
pid will end up "controversial" as a source of truth for process identification. because pids rotate, you could get 17895 three times today.. there are some creative/interesting alternatives here (most not cross-platform) but I think similar to host identification, we should dig a bit on the goal of process identification and where this hooks in and if/how we should apply multiple methodologies. 

most systems will leverage PID on the defensive side but some will not and we can expect this to be a sticking point with blue teams and even some IT.

lzimm replied:
re PIDs rotating: absolutely.

the key here though isn't to try and make things unique out of the box, it's just to capture what we can about the actual execution so that there's enough context for YOU to do the identification after the fact (using a combination of different factors, like timestamps and hostnames and etc) and triangulate on your needles inside your haystacks in whatever way makes the most sense.

because ultimately: commands, when run, DO have PIDs, and that part should be completely uncontroversial. i'd also like to capture the entire process hierarchy (of parent commands all the way up to the root), but that might be a bit of a stretch, and would settle for the parent process (ie: pneuma) PID as a way to, with a reasonable degree of accuracy, eliminate ambiguity when trying to pin things down (which should map fairly elegantly onto the process tree diagrams that falcon/elastic/etc. use to represent an attack as well).

whether or not a team wants to use that or not is up to them, and if they want to use some other crazy form of fingerprinting an actual process, then that's something that probably calls for a custom agent that tracks that through metadata on its own,

lzimm replied:
on that note:

instead of `pid`, what about:

- process:
  - id
  - parent_id

privateducky replied:
yep; that's pretty logical. allows people to build out that way

privateducky reviewed (changes requested):
overall, really great first pass. bunch of questions/comments... in general, I think we should:
- aim to make a little less flat, categorizing areas into objects like "host identification" or "process" or "classification"
- use (even more) generic language used so it can intend to be swapped out but maintain the purpose.
- highlight in the doc the agent-operator and hq-outpost properties. 
- determine which areas are in our domain of a link object and which can/should hook into other systems (like LDAP or asset inventory).

lzimm commented on docs/operator/advanced/links.md:
```diff
@@ -0,0 +1,49 @@
+- **payload**: if your TTP required a payload as part of its request, we store the specific payload name and version used.
+- **request**: this is the raw command body that got executed by the agent's executor, after all open variables have been filled by Operator, as it determined how to run your chain.
+- **response**: this is the result of running your command. if your command provides incremental results, you'll receive multiple links that belong to the same `operation` with the same `request` body, but have different `timelines` and responses. we capture:
+	- `output`: the regular output stream (stdout) of the request, which is where data gets parsed out of for downstream commands.
```
well, processes can still emit on stderr without failing completely, and vice versa (emitting on stdout, without succeeding). in any world though, it's never going to be harder to merge the two results (stdout/stderr) than it will be to disentangle them (ie: separating what out vs err was from a merged response) after the fact, which goes back to the whole "don't throw away context" thing.

lzimm commented on docs/operator/advanced/links.md:
```diff
@@ -0,0 +1,49 @@
+- **status**: the status code of the request.
+- **pid**: the process ID of the executed task.
+- **timeline**: if you're trying to correlate tasks that got executed by operator with the raw results you see in your logs, one of the most important heuristics you'll depend on is the time that each event was recorded to try to tell a story of what corresponds with what. however, since timestamps as a single moment don't capture the fact that instructions can take several milliseconds to hours  ...
+	- `start`: the moment before the agent started executing your command.
```
hahaha, yeah, so, there's a lot of stuff that goes on with the timestamp thing. at the end of the day though, what really matters is:

- the events that make it to the SIEM/EDR/etc. are all going to have the LOCAL time on the machine attached to them
- as long as our links capture time using the same relative frame, network time drift isn't that big of an issue (when it comes to correlating raw events, which is the main goal here), it's only a concern when it comes to trying to tell a story of the total order of things that happened at a higher level (which you can use the operator-side timestamps that i realized that we actually DO need to shove back in—ie: `timeline.generated`, `timeline.sent`, `timeline.received`—to do, simply by changing your reference frame)

the offset thing is just so we have a way of picking whether we want to make the time we display in the results themselves relative to your current timezone or the agent's timezone, inside operator, and i'm hesitant to put that inside the link itself simply because it's such a speculative value.

lzimm commented:
alright @privateducky , i've taken a stab at reformatting and trying to still keep top-level the actual task/result data itself, but shifted a lot of things inside separate buckets, which (though nothing final on the actual names themselves) feels kinda right to me.

- `origin` for what actually "initiated" the task: the user's account (which in the future would be the natural place to resolve the "license_key" or whatever value we end up needing there), the operation (and the possible chain that "drove" that), and the ttp
- `host` for where the task was executed: the agent, the hostname, the username, the mac address, and private IP (omitting the public IP as that feels a bit "extra", with results you could triangulate from other directions).
- `request` for the actual request: command and payload (tracking for reasons mentioned above, not to mention the built in checksum might allow teams to match executed binary signatures directly inside an EDR)
- `response` for the actual response data: output, error, and status code
- `process` for the actual process that executed the task: both the id and parent
- `timeline` for the actual execution timeline: generated, sent, received allow you to normalize tasks across agents from the reference frame of the co-ordinator, started and finished let you frame things relative to the events that actually happen on the system, between those two co-ordinate systems we should be able to satisfy our causality constraints.
- `facts` the facts in and out, as before
- `metadata` for auxiliary information and alternate implementations.

privateducky reviewed (commented):
Starting to come together nicely. 

So if I'm grabbing it right, you have these buckets (aside from id):
- host
- process
- request
- response
- timeline
- facts
- metadata

thinking about our ability to draw conclusions from these buckets:
1. is there any bucket of information naturally missing? 
2. thinking about criss-crossing of queries between sub-properties within these buckets, is there any reason this object structure would cause us issues?

I'll think these over too!

lzimm commented:
yup, those buckets are the core of what i could think of, with subtle tricks wrapped up inside each, especially the ones that augment data from multiple different sources.

in terms of missing, it really depends on what our hopes and goals are, i had a conversation with robert earlier where he relayed back from octavia for instance how they'd like to have "tags" one each link being published as well, but that starts getting us on another very quick slippery slope, where not only do we have to consider where we draw the line on what we seemingly arbitrarily decide to include vs. not include on each ttp, but also have to think, up front, even more, about how tags are actually used when authoring the ttp itself, in a way that might be shifting responsibility way too far into a bunch of different directions. so i'd pull back on that, and keep anything at that level inside the purview of reports that we can drive and host explicitly, with the ability to join on all of our TTP metadata, for instance, out of band.

additionally, there's the chains themselves, and what each link is connected to, and in what way. there might be an argument to be made that we should project into each link some sort of representation of the other links its connected to, but that, too, seems remarkably, speculative, to use a word, and something that you should be able to assemble (at least by hand) from the operation ids and drawing lines by grouping links on their ttp and dropping them on a graph with time on the x axis.

which brings me to the issues: the buckets themselves, as long as the platforms in question have a way of inspecting JSON objects and drawing appropriate joins on them, then in theory, no. in practice, however, that's really going to depend on the capabilities of the platform in question—their ergonomics for certain types of queries can make certain things "cumbersome" at times.

lzimm commented:
ohh, and on the missing point, the main other thing i think about is in the context of an EDR: if we're getting trampled by an EDR, or running beside an EDR and watching our own links go right through, it might be nice for an agent to be able to query the EDR on the status of each process it runs to collect that as part of the "status". but since that's also very speculative right now, if and when that might happen, at least there's room for it in the metadata field on an agent specifically designed to play battleship right up next to crowdstrike, etc.

privateducky reviewed (commented):
Keeping this open through the next sprint, so I can wrap my head around these changes alongside the dashboard layout ticket we have planned (both were originally in the same sprint because they go hand-in-hand). Definitely don't want to modify anything too heavily after we go through the effort to update/test.

lzimm commented:
yeah. i made some last minor adjustments the other day just adding chain/ttp names to links to make things more accessible to people without having to do a bunch of crazy/impractical joins all over the place if they want to roll up some very easy summary information, but i feel like that should be the last step. i wanna relax the schema on HQ today so i can push this new link format into an outpost and test things to really refine things based on how they feel in hand though, but hopefully that shouldn't be too much of an issue.

lzimm commented:
alright @privateducky, this should be good to go. removed a buncha things and moved a few other things around.

corresponding PR(s) for operator:

- https://github.com/preludeorg/operator/pull/3629 <-- base PR that gets the new format in
- https://github.com/preludeorg/operator/pull/3705 <-- PR on top of that one that makes changes to schedules/etc. in order to get chains forwarded all the way to results

privateducky commented on docs/operator/advanced/links.md:
```diff
@@ -0,0 +1,62 @@
+#### Attributes
+
+- `id`: every link has a unique id used to differentiate it from other executions, which may even share almost every other attribute in common. *(source: Operator)*
+- `origin`: information about the origin of the execution.
```
I'm good with this doc/plan - with the exception of the origin block. I think it goes too deep, too quick. I'd rather see some of these properties flatted, so at top level we have: tag (string), ttp (string) and chain (string). reasons:

- origin as a category is too generic; over time I think this will become a sticking point of what goes in. over time, we can evaluate these characteristics a bit better (no reason to rush).
- I don't want to create a bastardized ttp or chain object - yet. maybe we'll need to, but I'm not convinced. once you start doing this to a data model, it becomes too easy to add more and more until you've completely denormalized the object. maybe that's what we'll want - but I think we can keep an eye on it over the course of our testing this and next PSI. 
- the matching PR to this doc page is too complex for me to review in one swing. the changes I'm proposing in the origin section are to keep these props as is today, which means it can shrink our overall change set to make the PR easier to merge (and less risk).

lzimm replied:
so, i'm totally open to moving things out of origin, the extra nesting was definitely "one of these things not like the other"... that said though, we had a discussion around account emails, rather than just tags, and i think the "tag" is so generic that nesting it inside of a top level account object can help keep things a lot more legible for people.

outside of that, the executor which i've currently attached to the ttp as well. this can work multiple ways, but i wouldn't want to lose the executor as it seems like it could be valuable context (which we nuke if we just make the ttp a flat string)

lzimm replied:
flattened things a bit, but how does that look with the nested account object + ttp object (to contain the executor—we could clearly move that top level, but it starts to feel very "small" compared to the other things we're tracking at the object root)

lzimm replied:
nevermind, flattened executor out of ttp too, this keeps things symmetrical with the beacon object which i'm okay with. only thing left nested (out of origin) is `account`.

privateducky commented on docs/operator/advanced/links.md:
```diff
@@ -0,0 +1,58 @@
+#### Attributes
+
+- `id`: every link has a unique id used to differentiate it from other executions, which may even share almost every other attribute in common. *(source: Operator)*
+- `account`: we collect the account details of the user originating each request, this is an object with the following attributes:
```
I don't want email in our initial property set, primarily to keep PII out of logs/result sets. 

from an "account" perspective, there are 3 types of accounts to account for (get it?):
1. operator account. useful for tying back to license, member info, etc. on our end.
2. IAM/SSO at the organization itself
3. account/inventory of the machine itself.

For 1, we can leverage our tag if we need to look up additional info (like email, license, created date,..).

For 2, we'll end up hooking into several common auth mechanisms (a known set/quantity at least). For these, I actually bet we end up replacing our 'tag' value with ones that match user IDs in these systems.

For 3, we're mostly solving this through the host object in this doc. 

Because I feel that our tag will == the user ID mechanism at an organization, I think storing tag at the top-level of the link makes most sense. I could also see tag being dropped under the host object, as an identifier for the particular host. but I'd veer toward the former because many hosts either won't have an "owner" (like a server) and workstations are also prone to recycling in an org.

lzimm replied:
i'm totally fine abandoning the `email` there (and was only stubbing that in as a placeholder, looking ahead to whatever other "creds" we'd want to stuff in there in the future).

with that said though, i really i think the name `tag` alone is kind of meaningless to most people (and i think we should be attempting to model these things in way that have a fairly low amount of mental friction that people don't have to keep recycling their memory every time they read a term that doesn't trigger an intuitive image of what it means).

and finally, even though i would strongly prefer to keep these things nested, because even though i agree that we'll be tying our "tags" to users at the org at some point, i don't think we can really say one way or another if the full account saturation will need to happen at our end (ie: if we'll need to track identifiers back into their systems as part of our organizational models that need to get filled in en route through HQ) or on their end (ie: if they even have a place to shove our user tags somewhere, along side their existing user management systems, in a convenient way, so they can do the lookups wherever they need to), and my intent in laying this out was to account for that possibility in a much more future proof way. but because of the way everything else got flattened, it doesn't really make sense now to keep this one as the only one that isn't, but if the goal is to future proof our link formats, i can see this one in particular being one of the main ones that's likely prone to "adaptation".

taking both of those last notes in mind though, even though i'd rather label our `tag` as `account` because i think its more meaningful to people, since i'd rather add a new `account` structure in the future rather than suddenly change the datatype OF that `account` field from a uuid into a nested structure, `tag` it is...

privateducky commented on docs/operator/advanced/links.md:
```diff
@@ -0,0 +1,58 @@
+- `operation`: every link belongs to an operation, which is the total set of every TTP, and all of the requests it generates, when when you decide to run a chain. this is the id of the operation itself. *(source: Operator)*
+- `chain`: if the operation was spawned by a chain, the id of the chain itself. *(source: Operator)*
+- `ttp`: the TTP id of the attack being executed. *(source: Operator)*
+- `executor`: since each TTP can be executed using a variety of different executors (ie: bash, powershell, python, etc.), we store the executor selected to run the individual instruction to help you distinguish that too. *(source: Operator)*
```
executor feels like an outlier at the top-level. I can see why we'd need it, but I'd bucket it under "request", as that seems natural.

lzimm replied:
duh, i don't know how i missed that one.

privateducky commented:
as promised, new round of thoughts!

the biggest thing I'm weighing right now is the appropriate level of normalization. I think what we have in this proposed link format is the ideal level of normalization for a "regular" database (a SQL or non-SQL that is importantly wrapped by our own business logic, so we control the saturation of the data with other internal/external data sources, like TTPs or asset inventory).

because we're routing data into SIEM or other highly-optimized for streaming data DBs, I think we should denormalize the data that is an internal-to-us data structure and keep normalization for external-to-us data (note: this comment is much longer than the actual change to the object I'm weighing here!). I was originally hesitant to go this route because I always like to carefully wade into data duplication instead of making it my default position.

what this would look like in practice (largely a review, as this all exists, once your agent/beacon PR is in):
- a beacon object would be a subset of the agent object inside Operator.
    - the beacon would continue containing host identifiers, operational configuration props, and a list of links. 
    - the agent object would be split into two; the beacon-related properties and the Operator/C2 specific properties (like history)  
- a link object on the beacon would look like it does today, with tight information that was necessary for the agent <--> operator transaction. 
    - when Operator receives a beacon, any agent-level properties would get updated on the receiving agent object.
    - each link on the beacon would be saturated per this doc (some of the saturation occurring before the instruction is sent to the agent) and sent to HQ and routed to Outpost.

what adding denormalization would mean in light of this doc. we'd:
- store the full TTP per link, which would replace the 'request' property.
- store the list of included TTPs in the operation property 
- store the full chain object in the chain property.
- the tag property would remain as-is, being a tie to a future (unknown) data source.
- the host property would remain as-is but I'd look at it as the section of the link we're storing properties from the corresponding agent object.

whatcha think?

lzimm commented:
hah, so, welcome back to the dark side @privateducky ...

anyways, though i definitely think in order to be useful, we need to store more than just UUIDs, i wouldn't want to store the full object because that locks in our TTP/Chain schemas really tightly all at once too (and I'm not sure how much value it REALLY is to store all the different commands you DIDN'T run on each link, as much of a curiosity it might be, for instance, either).

personally i'd opt for being more "minimum viably legible", which boils down to instead of

```
chain: uuid
ttp: uuid
```

going back to something more like what i originally had:

```
chain:
  - id
  - name
ttp:
  - id
  - name
  - tactic
  - technique
```

which should satisfy most of the up front use cases that i think we're really going to be confronted with out of the box (though if we want to append some different type of taxonomy information on that, that works too).

one thing, connected back to your list of included TTPs, i could be persuaded to putting that in either the operation or chain pretty easily, but since you can't really join across links inside a SIEM very easily without a bunch of custom stuff anyways, i'm not sure how much value there is there that you can't already compute from just using the operation and working backwards from there instead.

UNLESS we're talking about being able to check for ttps that DIDN'T get run as part of an operation/chain, THAT could be a very different story, but on that end, maybe just a (ttp, not link, because we can't compute that until the fact amplification happens live) COUNT, instead of the full list of (again, ttp, not link) IDs?

privateducky commented:
where I'm stuck is, I'm trying to go one of two extremes: normalized or denormalized, and trying to avoid the in between version which requires creating bastardized/convenient/adhoc forms of our existing objects (primarily chain, ttp and operation). 

if we venture into that middle land, it becomes a game of constant shoving props into the link that we need along the way. one that comes to mind is, a query where I want to see the description of each TTP that was executed:
1. denormalized (full TTP in the link) allows me to find it quickly.
2. normalized (just a TTP ID) means I need to JOIN my query with another source
3. bastardized version means I'll just dump the description into the link when I decide I want it (creating inconsistent data if I wasn't doing this before).

the final solution may be a version between 1 and 2, where we add a function in our Operator objects similar to the ".display" we use in HQ to serialize responses from the API. If we created a ttp.display, chain.display and operation.display, we could formalize structures for these objects and attach them when publishing links. 

thoughts?

lzimm commented:
@privateducky operator side implementation of what we talked about earlier is here: https://github.com/preludeorg/operator/pull/3784 — keeps things normalized on local disk, but inflates full objects when we push to HQ

privateducky reviewed (changes requested):
drop chain + process.parent (for now) and this one will line up well with the changes in your PR and be a good win for this sprint.

lzimm commented:
> drop chain + process.parent (for now) and this one will line up well with the changes in your PR and be a good win for this sprint.

done!

lzimm commented:
woo!
